src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/PreferDictionaryTryGetValueAnalyzer.cs

```diff
@@ -0,0 +1,145 @@
+            context.ReportDiagnostic(Diagnostic.Create(ContainsKeyRule, containsKeyInvocation.Syntax.GetLocation(), additionalLocations));
+        }
+
+        private static bool TryGetContainsKeyGuard(IConditionalOperation conditionalOperation, [NotNullWhen(true)] out IInvocationOperation? containsKeyInvocation)
+        {
+            containsKeyInvocation = conditionalOperation.Condition as IInvocationOperation ?? FindContainsKeyInvocation(conditionalOperation.Condition);
+            if (containsKeyInvocation is not null && containsKeyInvocation.TargetMethod.Name == ContainsKeyMethodName)
+            {
+                return true;
+            }
+
+            return false;
+        }
+
+        private static IInvocationOperation? FindContainsKeyInvocation(IOperation baseOperation)
+        {
+            return baseOperation switch
+            {
+                IInvocationOperation i when i.TargetMethod.Name == ContainsKeyMethodName => i,
+                IBinaryOperation { OperatorKind: BinaryOperatorKind.ConditionalAnd or BinaryOperatorKind.ConditionalOr } b =>
+                    FindContainsKeyInvocation(b.LeftOperand) ?? FindContainsKeyInvocation(b.RightOperand),
+                _ => null
+            };
+        }
+
+        private static bool DictionaryEntryIsModified(IPropertyReferenceOperation dictionaryAccess, IBlockOperation blockOperation)
+        {
+            return blockOperation.Operations.OfType<IExpressionStatementOperation>().Any(o =>
+                o.Operation is IAssignmentOperation { Target: IPropertyReferenceOperation reference } && reference.Property.Equals(dictionaryAccess.Property, SymbolEqualityComparer.Default));
+        }
+
+        private static bool IsDictionaryAccess(IPropertyReferenceOperation propertyReference, INamedTypeSymbol dictionaryType)
+        {
+            return propertyReference.Property.IsIndexer && IsDictionaryType(propertyReference.Property.ContainingType, dictionaryType) &&
+                   (propertyReference.Property.OriginalDefinition.Name == IndexerName || propertyReference.Language == "Visual Basic" && propertyReference.Property.OriginalDefinition.Name == IndexerNameVb);
+        }
+
+        private static bool TryGetParentConditionalOperation(IOperation derivedOperation, [NotNullWhen(true)] out IConditionalOperation? conditionalOperation)
+        {
+            conditionalOperation = null;
+            do
+            {
+                if (derivedOperation.Parent is IConditionalOperation c)
+                {
+                    conditionalOperation = c;
+
+                    return true;
+                }
+
+                derivedOperation = derivedOperation.Parent;
+            } while (derivedOperation.Parent != null);
+
+            return false;
+        }
+
+        private static bool IsDictionaryType(INamedTypeSymbol derived, ISymbol dictionaryType)
+        {
+            var constructedDictionaryType = derived.GetBaseTypesAndThis()
```

Comment:
I remember something about the analyzer needing to find derived types as well, but of course can't find any evidence of that now 😅 
Would it be a bad thing to catch derived types as well?

Reply:
I think because they could have different implementation than .Net and might not cause the issue we trying to fix with the analyzer. Though in this case I doubt that the implementation of `ContainsKey+indexer` could avoid double lookup.

Reply:
Good point. So should I remove it, just to be safe?

Reply:
Okay, I'll get on that. This also means that we won't flag a `ContainsKey` on an `IDictionary`, right? Most likely the implementation is a .NET `Dictionary`, however this is not guaranteed. This is unfortunate, since most people (myself included) use `IDictionary` when declaring a `Dictionary`.

Reply:
Good point, most cases the derived types implementation would have same issue, so lets keep checking derived types

Reply:
Done. I made some changes to increase symbol comparisons, however I am still running into issues when comparing i.e. the `ConcurrentDictionary.ContainsKey` method and the `IDictionary.ContainsKey` methods. Even their generic arguments are symbolically different, so the only way I found was to compare signature elements (return type, name and parameters) via string comparison. Is there any way to compare an implemented method to it's interface version more reliably?

Reply:
I would use: https://github.com/dotnet/roslyn-analyzers/blob/b2c2cbef531dd778a6b4020a7f9fdda9d010923c/src/Utilities/Compiler/Extensions/ISymbolExtensions.cs#L525 but seems it is not working for this symbols cc @mavasani